### ISSUE TITLE
Add ability to include and filter via detail fields on pages' list endpoints

### ISSUE THREAD

**Issue by berzi** [wagtail/wagtail#11404]
### Is your proposal related to a problem?

There is currently no way to "find" pages via fields (the typical example being [by slug](https://stackoverflow.com/questions/50765291/returning-all-fields-with-wagtail-api-from-page-slug) and retrieve the values for "detail" fields of the model in one go: you have to use the list endpoint to search, retrieve the ID, then use the ID to call the detail endpoint and retrieve the data. This is extremely wasteful considering how common this operation is, and I'm not aware of any better alternative.

### Describe the solution you'd like

As alleged¹ in the [answer to the aforelinked SO question](https://stackoverflow.com/a/50779199/8599834) citing a Slack response, Wagtail itself runs into this issue and solves it by expanding the PagesAPIViewSet. Instead of forcing users of Wagtail's API to do this themselves for every project, a generic solution could be integrated into PagesAPIViewSet itself.

It would be helpful if any user who already wrote a solution to this for themselves or is already familiar with that part of Wagtail's code would share their code in a PR so that everyone can use it.

The features needed would essentially be:
- When a `fields` query parameter is supplied to the list endpoint, lookup the fields in the `specific()` model.
- When an individual field is supplied in the query parameters with a value to match it (e.g. `?slug=my-slug`), the value is looked up in the `specific()` models instead of the reduced list version of it.

¹ The links to the old Wagtail docs and code are now dead.

**Comment by berzi** [comment]
As a note for anyone wanting to attempt this, I solved it in my own project with the following:

```py
# The wagtail_router points to this class instead of PagesAPIViewSet.
class BlogPagesAPIViewSet(PagesAPIViewSet):
    # BlogPost is a Page subclass. The default PagesAPIViewSet.model is Page itself,
    #  so fields that are specific to BlogPost are not found by the methods that use
    #  self.model to check for its fields.
    model = BlogPost
```

I plan to have separate API routes (hence separate ViewSets) for different kinds of pages, but a general solution would need to somehow have the specific model in place of `PagesAPIViewSet.model`, or get it at runtime wherever relevant instead/on top of using `PagesAPIViewSet.model`.

**Comment by zerolab** [comment]
This gets a bit more complicated in that we'd need to iterate through all Page models and use [Q objects](https://docs.djangoproject.com/en/5.0/topics/db/queries/#complex-lookups-with-q-objects), but only if the field is defined on the specific Page model.

Your separate API routes approch is the right way, IMHO and perhaps what we need to do is add a recipe to the docs, rather than make `PagesAPIViewSet` do everything

**Comment by berzi** [comment]
I agree, since it's a simple solution anyway, but I can't shake off the thought that having the detail and list endpoints behave so differently violates the {principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) (having been baffled by it myself very recently when I found out).

A documentation addition is certainly a sufficient "solution" in the short term, but if a real one can come baked into Wagtail without considerable drawbacks, I think it would be nice to have it. In my case, I had to subclass and separate the viewset for [other reasons](https://github.com/wagtail/wagtail/issues/7497#issuecomment-1874291366) too, but the typical user probably needs this feature even on its own (I find it very likely that someone using Wagtail headlessly also wants infinite scrolling, for example).
